File: lib/ts/src/protocol.ts
export enum MsgType {
    Call = 1,
    Response = 2,
}

export type Vals = any[];

export interface CallMessage {
    type: MsgType.Call,
    id: number,
    method: string;
    args: Vals;
}

export interface ResponseMessage {
    type: MsgType.Response,
    id: number,
    result?: Vals;
    error?: string;
}

export type Message = CallMessage | ResponseMessage;

export interface CallResult {
    result: Vals;
    error: Error | null;
}

File: lib/golang/protocol.go
package golang

const ipcSocketArg = "--ipc-socket"
const maxMessageLength = 1 << 30 // 1 GB
const defaultAcceptTimeout = 10 // seconds

type MsgType int

type Vals []any

const (
	MsgCall     MsgType = 1
	MsgResponse MsgType = 2
)

type Message struct {
	Type   MsgType `json:"type"`
	Id     int64   `json:"id"`
	Method string  `json:"method"`
	Args   Vals    `json:"args"`
	Result Vals    `json:"result"`
	Error  string  `json:"error"`
}
